model_signing/serialization/dfs.py

```diff
@@ -0,0 +1,303 @@
+                digest = self._file_hasher.compute()
+                hasher.update(digest.digest_value)
+            else:
+                header = _build_header(entry_name=child.name, entry_type="dir")
+                hasher.update(header)
+                digest = self._dfs(child)
+                hasher.update(digest.digest_value)
+
+        return hasher.compute()
+
+
+# Define type aliases for the ShardedDFSSerializer class below.
+_SizedPath: TypeAlias = tuple[pathlib.PurePath, str, int]
+_ShardSignTask: TypeAlias = tuple[pathlib.PurePath, str, int, int]
+
+
+def _endpoints(step: int, end: int) -> Iterable[int]:
+    """Yields numbers from `step` to `end` inclusive, spaced by `step`.
+
+    Last value is always equal to `end`, even when `end` is not a multiple of
+    `step`. There is always a value returned.
+
+    Examples:
+    ```python
+    >>> list(_endpoints(2, 8))
+    [2, 4, 6, 8]
+    >>> list(_endpoints(2, 9))
+    [2, 4, 6, 8, 9]
+    >>> list(_endpoints(2, 2))
+    [2]
+    """
+    for value in range(step, end, step):
+        yield value
+    yield end
+
+
+class ShardedDFSSerializer(serialization.Serializer):
+    """DFSSerializer that uses a sharded hash engine to exploit parallelism."""
+
+    def __init__(
+        self,
+        file_hasher_factory: Callable[
+            [pathlib.Path, int, int], file.ShardedFileHasher
+        ],
+        merge_hasher: hashing.StreamingHashEngine,
+        max_workers: int | None = None,
+    ):
+        """Initializes an instance to hash a file with a specific `HashEngine`.
+
+        Args:
+            hasher_factory: A callable to build the hash engine used to hash
+              every shard of the files in the model. Because each shard is
+              processed in parallel, every thread needs to call the factory to
+              start hashing. The arguments are the file, and the endpoints of
+              the shard.
+            merge_hasher: A `hashing.StreamingHashEngine` instance used to merge
+              individual file digests to compute an aggregate digest.
+            max_workers: Maximum number of workers to use in parallel. Default
+              is to defer to the `concurent.futures` library.
+        """
+        self._file_hasher_factory = file_hasher_factory
+        self._merge_hasher = merge_hasher
+        self._max_workers = max_workers
+
+        # Precompute some private values only once by using a mock file hasher.
+        # None of the arguments used to build the hasher are used.
+        hasher = file_hasher_factory(pathlib.Path(), 0, 1)
+        self._shard_size = hasher.shard_size
+
+    @override
+    def serialize(self, model_path: pathlib.Path) -> manifest.Manifest:
+        # TODO: #196 - Add checks to exclude symlinks if desired
+        _check_file_or_directory(model_path)
+
+        if model_path.is_file():
+            entries = [model_path]
+        else:
+            # TODO: When Python3.12 is the minimum supported version, this can
```

Comment:
nit: create a tracking issue

Reply:
Good idea, but it won't be for the v1 milestone. Made #200